domain/infosync/info.go

```diff
@@ -451,6 +460,91 @@ func (is *InfoSyncer) updateTopologyAliveness(ctx context.Context) error {
 		clientv3.WithLease(is.topologySession.Lease()))
 }
 
+// GetPrometheusAddr gets prometheus Address
+func GetPrometheusAddr() (string, error) {
+	is, err := getGlobalInfoSyncer()
+	if err != nil {
+		return "", err
+	}
+
+	// if the cache of prometheusAddr is over 5s, update the prometheusAddr
+	if time.Since(is.modifyTime) < TablePrometheusCacheExpiry {
+		return is.prometheusAddr, nil
+	}
+	is.prometheusAddr = ""
+	return is.getPrometheusAddr()
+}
+
+type prometheus struct {
+	IP         string `json:"ip"`
+	BinaryPath string `json:"binary_path"`
+	Port       int    `json:"port"`
+}
+
+type metricStorage struct {
+	PdServer struct {
+		MetricStorage string `json:"metric-storage"`
+	} `json:"pd-server"`
+}
+
+func (is *InfoSyncer) getPrometheusAddr() (string, error) {
+	// Get PD servers info.
+	pdAddrs := is.etcdCli.Endpoints()
+	if len(pdAddrs) < 0 {
+		return "", errors.New("pd unavailable")
+	}
+	var res string
+
+	// get prometheus address from pdApi
+	var url string
+	if strings.HasPrefix(pdAddrs[0], "http://") {
+		url = fmt.Sprintf("%s%s", pdAddrs[0], pdapi.Config)
+	} else {
+		url = fmt.Sprintf("http://%s%s", pdAddrs[0], pdapi.Config)
+	}
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	var metricStorage metricStorage
+	dec := json.NewDecoder(resp.Body)
+	err = dec.Decode(&metricStorage)
+	if err != nil {
+		return "", err
+	}
+	res = metricStorage.PdServer.MetricStorage
+
+	// get prometheus address from etcdApi
+	if res == "" {
+		values, err := is.get(TopologyPrometheus)
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		var prometheus prometheus
+		err = json.Unmarshal([]byte(values), &prometheus)
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		res = fmt.Sprintf("http://%s:%v", prometheus.IP, strconv.Itoa(prometheus.Port))
+	}
+	is.prometheusAddr = res
+	is.modifyTime = time.Now()
+	return res, nil
+}
+
+func (is *InfoSyncer) get(k string) (string, error) {
```

Comment:
How about rename to `getPrometheusAddrFromEtcd()`

Reply:
great! done.